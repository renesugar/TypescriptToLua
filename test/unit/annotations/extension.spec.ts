import {
    extensionCannotConstruct,
    extensionCannotExtend,
    extensionInvalidInstanceOf,
} from "../../../src/transformation/utils/diagnostics";
import * as util from "../../util";

test.each(["extension", "metaExtension"])("Class extends extension (%p)", extensionType => {
    util.testModule`
        declare class A {}
        /** @${extensionType} **/
        class B extends A {}
        class C extends B {}
    `.expectDiagnosticsToMatchSnapshot([extensionCannotExtend.code]);
});

test.each(["extension", "metaExtension"])("Class construct extension (%p)", extensionType => {
    util.testModule`
        declare class A {}
        /** @${extensionType} **/
        class B extends A {}
        const b = new B();
    `.expectDiagnosticsToMatchSnapshot([extensionCannotConstruct.code]);
});

test.each(["extension", "metaExtension"])("instanceof extension (%p)", extensionType => {
    util.testModule`
        declare class A {}
        /** @${extensionType} **/
        class B extends A {}
        declare const foo: any;
        const result = foo instanceof B;
    `.expectDiagnosticsToMatchSnapshot([extensionInvalidInstanceOf.code]);
});
