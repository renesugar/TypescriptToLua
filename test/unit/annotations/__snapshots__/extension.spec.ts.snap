// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Class construct extension ("extension"): code 1`] = `
"require(\\"lualib_bundle\\");
b = __TS__New(B)"
`;

exports[`Class construct extension ("extension"): diagnostics 1`] = `"main.ts(5,19): error TSTL: Cannot construct classes with '@extension' or '@metaExtension' annotation."`;

exports[`Class construct extension ("metaExtension"): code 1`] = `
"require(\\"lualib_bundle\\");
local __meta__A = debug.getregistry().A
b = __TS__New(B)"
`;

exports[`Class construct extension ("metaExtension"): diagnostics 1`] = `"main.ts(5,19): error TSTL: Cannot construct classes with '@extension' or '@metaExtension' annotation."`;

exports[`Class extends extension ("extension"): code 1`] = `
"require(\\"lualib_bundle\\");
C = __TS__Class()
C.name = \\"C\\"
__TS__ClassExtends(C, B)"
`;

exports[`Class extends extension ("extension"): diagnostics 1`] = `"main.ts(5,9): error TSTL: Cannot extend classes with '@extension' or '@metaExtension' annotation."`;

exports[`Class extends extension ("metaExtension"): code 1`] = `
"require(\\"lualib_bundle\\");
local __meta__A = debug.getregistry().A
C = __TS__Class()
C.name = \\"C\\"
__TS__ClassExtends(C, B)"
`;

exports[`Class extends extension ("metaExtension"): diagnostics 1`] = `"main.ts(5,9): error TSTL: Cannot extend classes with '@extension' or '@metaExtension' annotation."`;

exports[`instanceof extension ("extension"): code 1`] = `
"require(\\"lualib_bundle\\");
result = __TS__InstanceOf(foo, B)"
`;

exports[`instanceof extension ("extension"): diagnostics 1`] = `"main.ts(6,24): error TSTL: Cannot use instanceof on classes with '@extension' or '@metaExtension' annotation."`;

exports[`instanceof extension ("metaExtension"): code 1`] = `
"require(\\"lualib_bundle\\");
local __meta__A = debug.getregistry().A
result = __TS__InstanceOf(foo, B)"
`;

exports[`instanceof extension ("metaExtension"): diagnostics 1`] = `"main.ts(6,24): error TSTL: Cannot use instanceof on classes with '@extension' or '@metaExtension' annotation."`;
